### ISSUE TITLE
kube-scheduler doesn't respect ResourceExhausted

### ISSUE THREAD

**Issue by mrpre** [kubernetes-csi/external-provisioner#697]
* Environment
`v1.22.3`
* Deployment 
1. `CSIDriver` is deployed with 
```
  attachRequired: false
  podInfoOnMount: true
  storageCapacity: true
```
2. `StorageClass` is deployed with
```
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
```
2. `csi-provisioner` with plugin is deployed using `Daemonset` like this
```
image: csi-provisioner:v3.1.0
args:
  - "--node-deployment"
  - "--csi-address=$(ADDRESS)"
  - "--v=5"
  - "--feature-gates=Topology=True"
  - "--enable-capacity"
  - "--extra-create-metadata=true"
image: myplugin-local
args:
  - "--csi-address=$(ADDRESS)"
  - "--v=5"
  ...
```
It seems thant I have followed all the [requirement](https://github.com/kubernetes-csi/external-provisioner#capacity-support) such as `environment variables` and `command line flags`  

* current node storage capacity
Following `csistoragecapacities` is  the result after I deployed a  StatefulSet with `replicas: 5` and `storage: 100G` 
```
$kubectl describe csistoragecapacities
Name:                 csisc-7qmsw
Namespace:            default
Labels:               csi.storage.k8s.io/drivername=pmemcsi.csi.alibabacloud.com
                      csi.storage.k8s.io/managed-by=external-provisioner-cn-hangzhou.10.0.0.103
Capacity:             49326344Ki
Maximum Volume Size:  24663172Ki

Name:                 csisc-g5pjt
Namespace:            default
Labels:               csi.storage.k8s.io/drivername=pmemcsi.csi.alibabacloud.com
                      csi.storage.k8s.io/managed-by=external-provisioner-cn-hangzhou.10.0.0.105
Capacity:             154183944Ki
Maximum Volume Size:  129520772Ki

Name:                 csisc-r6rqj
Namespace:            default
Labels:               csi.storage.k8s.io/drivername=pmemcsi.csi.alibabacloud.com
                      csi.storage.k8s.io/managed-by=external-provisioner-cn-hangzhou.10.0.0.104
Capacity:             45128192Ki
Maximum Volume Size:  22625532Ki
```

when I deploy a new StatefulSet with `replicas: 1`and it's storage requests setting to `40G `, the storage should be provisioned by  node `cn-hangzhou.10.0.0.105`.

But `CreateVolume` is always called on node `cn-hangzhou.10.0.0.103` 
the log of `external-provisioner ` shows `supports topology = true, node selected true => may reschedule = true => state = Reschedule` which indicate the the return value is correctly
```
I0126 08:51:57.809074   48303 controller.go:1337] provision "default/pv-data4-stspv4-0" class "pmemcsi": started
I0126 08:51:57.809152   48303 connection.go:183] GRPC call: /csi.v1.Controller/CreateVolume
I0126 08:51:57.809183   48303 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv-data4-stspv4-0", UID:"dbf950c1-a941-429f-b7e7-999e15b94877", APIVersion:"v1", ResourceVersion:"64428", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/pv-data4-stspv4-0"
I0126 08:51:57.809158   48303 connection.go:184] GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"pmemcsi.csi.alibabacloud.com":"enable"}}],"requisite":[{"segments":{"pmemcsi.csi.alibabacloud.com":"enable"}}]},"capacity_range":{"required_bytes":42949672960},"name":"pvc-dbf950c1-a941-429f-b7e7-999e15b94877","parameters":{"csi.storage.k8s.io/pv/name":"pvc-dbf950c1-a941-429f-b7e7-999e15b94877","csi.storage.k8s.io/pvc/name":"pv-data4-stspv4-0","csi.storage.k8s.io/pvc/namespace":"default"},"volume_capabilities":[{"AccessType":{"Mount":{}},"access_mode":{"mode":1}}]}
I0126 08:51:57.816053   48303 connection.go:186] GRPC response: {}
I0126 08:51:57.816082   48303 connection.go:187] GRPC error: rpc error: code = ResourceExhausted desc = CreateVolume: get available path from [/hostmnt/pmem1 /hostmnt/pmem0] with error: available path not found in [/hostmnt/pmem1 /hostmnt/pmem0] may no enough space for request 42949672960 bytes
I0126 08:51:57.816094   48303 controller.go:764] CreateVolume failed, supports topology = true, node selected true => may reschedule = true => state = Reschedule: rpc error: code = ResourceExhausted desc = CreateVolume: get available path from [/hostmnt/pmem1 /hostmnt/pmem0] with error: available path not found in [/hostmnt/pmem1 /hostmnt/pmem0] may no enough space for request 42949672960 bytes
I0126 08:51:57.816104   48303 capacity.go:450] Capacity Controller: enqueuing {segment:0xc00054c408 storageClassName:pmemcsi} because of the storage class
I0126 08:51:57.816198   48303 capacity.go:565] Capacity Controller: refreshing {segment:0xc00054c408 storageClassName:pmemcsi}
I0126 08:51:57.816221   48303 connection.go:183] GRPC call: /csi.v1.Controller/GetCapacity
I0126 08:51:57.816225   48303 connection.go:184] GRPC request: {"accessible_topology":{"segments":{"pmemcsi.csi.alibabacloud.com":"enable"}},"volume_capabilities":[{"AccessType":{"Mount":null},"access_mode":{}}]}
I0126 08:51:57.816266   48303 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv-data4-stspv4-0", UID:"dbf950c1-a941-429f-b7e7-999e15b94877", APIVersion:"v1", ResourceVersion:"64428", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "pmemcsi": rpc error: code = ResourceExhausted desc = CreateVolume: get available path from [/hostmnt/pmem1 /hostmnt/pmem0] with error: available path not found in [/hostmnt/pmem1 /hostmnt/pmem0] may no enough space for request 42949672960 bytes
I0126 08:51:57.821105   48303 connection.go:186] GRPC response: {"available_capacity":50510176256,"maximum_volume_size":{"value":25255088128}}
I0126 08:51:57.821147   48303 connection.go:187] GRPC error: <nil>
I0126 08:51:57.821162   48303 capacity.go:659] Capacity Controller: updating csisc-7qmsw for {segment:0xc00054c408 storageClassName:pmemcsi}, new capacity 49326344Ki
I0126 08:51:57.825197   48303 capacity.go:664] Capacity Controller: updated csisc-7qmsw with new resource version 29731 for {segment:0xc00054c408 storageClassName:pmemcsi} with capacity 49326344Ki
I0126 08:51:57.825606   48303 controller.go:1429] provision "default/pv-data4-stspv4-0" class "pmemcsi": volume rescheduled because: failed to provision volume with StorageClass "pmemcsi": rpc error: code = ResourceExhausted desc = CreateVolume: get available path from [/hostmnt/pmem1 /hostmnt/pmem0] with error: available path not found in [/hostmnt/pmem1 /hostmnt/pmem0] may no enough space for request 42949672960 bytes
I0126 08:51:57.825619   48303 controller.go:1071] Stop provisioning, removing PVC dbf950c1-a941-429f-b7e7-999e15b94877 from claims in progress
I0126 08:51:57.825639   48303 controller.go:1299] should provision: checking node for PVC default/pv-data4-stspv4-0 with resource version 64428
I0126 08:51:57.825645   48303 controller.go:1301] should provision: done checking node for PVC default/pv-data4-stspv4-0 with resource version 64428: provision false, err <nil>
```  

BUT the kube-scheduler is always try to assume the node `cn-hangzhou.10.0.0.103`

`kube-scheduler` log: 
```
I0126 15:20:15.339068       1 leaderelection.go:278] successfully renewed lease kube-system/kube-scheduler
I0126 15:20:16.803551       1 eventhandlers.go:123] "Add event for unscheduled pod" pod="default/stspv4-0"
I0126 15:20:16.803605       1 scheduling_queue.go:928] "About to try and schedule pod" pod="default/stspv4-0"
I0126 15:20:16.803623       1 scheduler.go:516] "Attempting to schedule pod" pod="default/stspv4-0"
I0126 15:20:16.803786       1 scheduler_binder.go:742] PVC "default/pv-data4-stspv4-0" is not bound
I0126 15:20:16.803902       1 csi.go:188] "Persistent volume had no name for claim" PVC="default/pv-data4-stspv4-0"
I0126 15:20:16.803944       1 scheduler_binder.go:266] FindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.105"
I0126 15:20:16.803973       1 scheduler_binder.go:856] No matching volumes for Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0" on node "cn-hangzhou.10.0.0.105"
I0126 15:20:16.804021       1 scheduler_binder.go:920] Provisioning for 1 claims of pod "default/stspv4-0" that has no matching volumes on node "cn-hangzhou.10.0.0.105" ...
I0126 15:20:16.804067       1 csi.go:188] "Persistent volume had no name for claim" PVC="default/pv-data4-stspv4-0"
I0126 15:20:16.804100       1 scheduler_binder.go:266] FindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.103"
I0126 15:20:16.804116       1 scheduler_binder.go:856] No matching volumes for Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0" on node "cn-hangzhou.10.0.0.103"
I0126 15:20:16.804135       1 scheduler_binder.go:920] Provisioning for 1 claims of pod "default/stspv4-0" that has no matching volumes on node "cn-hangzhou.10.0.0.103" ...
I0126 15:20:16.804164       1 csi.go:188] "Persistent volume had no name for claim" PVC="default/pv-data4-stspv4-0"
I0126 15:20:16.804192       1 scheduler_binder.go:266] FindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.104"
I0126 15:20:16.804208       1 scheduler_binder.go:856] No matching volumes for Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0" on node "cn-hangzhou.10.0.0.104"
I0126 15:20:16.804238       1 scheduler_binder.go:920] Provisioning for 1 claims of pod "default/stspv4-0" that has no matching volumes on node "cn-hangzhou.10.0.0.104" ...

I0126 15:20:16.804561       1 scheduler_binder.go:378] AssumePodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.103"
I0126 15:20:16.804634       1 scheduler_binder.go:742] PVC "default/pv-data4-stspv4-0" is not bound
I0126 15:20:16.804657       1 scheduler_assume_cache.go:328] Assumed v1.PersistentVolumeClaim "default/pv-data4-stspv4-0", version 34774
I0126 15:20:16.804732       1 volume_binding.go:329] "Trying to bind volumes for pod" pod="default/stspv4-0"
I0126 15:20:16.804740       1 scheduler_binder.go:451] BindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.103"
I0126 15:20:16.804750       1 scheduler_binder.go:537] bindAPIUpdate: Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0"
I0126 15:20:16.811357       1 round_trippers.go:454] PUT https://10.0.0.100:6443/api/v1/namespaces/default/persistentvolumeclaims/pv-data4-stspv4-0 200 OK in 6 milliseconds
I0126 15:20:17.344001       1 round_trippers.go:454] GET https://10.0.0.100:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=5s 200 OK in 3 milliseconds
I0126 15:20:17.348735       1 round_trippers.go:454] PUT https://10.0.0.100:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=5s 200 OK in 4 milliseconds
I0126 15:20:17.348862       1 leaderelection.go:278] successfully renewed lease kube-system/kube-scheduler
I0126 15:20:17.551260       1 cache.go:732] Couldn't expire cache for pod default/stspv4-0. Binding is still in progress.
I0126 15:20:17.811992       1 volume_binding.go:332] "Failed to bind volumes for pod" pod="default/stspv4-0" err="binding volumes: provisioning failed for PVC \"pv-data4-stspv4-0\""
E0126 15:20:17.812031       1 framework.go:863] "Failed running PreBind plugin" err="binding volumes: provisioning failed for PVC \"pv-data4-stspv4-0\"" plugin="VolumeBinding" pod="default/stspv4-0"
I0126 15:20:17.812059       1 scheduler_assume_cache.go:342] Restored v1.PersistentVolumeClaim "default/pv-data4-stspv4-0"
E0126 15:20:17.812120       1 factory.go:397] "Error scheduling pod; retrying" err="running PreBind plugin \"VolumeBinding\": binding volumes: provisioning failed for PVC \"pv-data4-stspv4-0\"" pod="default/stspv4-0"
I0126 15:20:17.812219       1 scheduler.go:414] "Updating pod condition" pod="default/stspv4-0" conditionType=PodScheduled conditionStatus=False conditionReason="SchedulerError"
I0126 15:20:17.819926       1 round_trippers.go:454] PATCH https://10.0.0.100:6443/api/v1/namespaces/default/pods/stspv4-0/status 200 OK in 7 milliseconds
I0126 15:20:17.821008       1 round_trippers.go:454] POST https://10.0.0.100:6443/apis/events.k8s.io/v1/namespaces/default/events 201 Created in 8 milliseconds




I0126 15:20:19.091397       1 scheduling_queue.go:928] "About to try and schedule pod" pod="default/stspv4-0"
I0126 15:20:19.091432       1 scheduler.go:516] "Attempting to schedule pod" pod="default/stspv4-0"
I0126 15:20:19.091543       1 scheduler_binder.go:742] PVC "default/pv-data4-stspv4-0" is not bound
I0126 15:20:19.091651       1 csi.go:188] "Persistent volume had no name for claim" PVC="default/pv-data4-stspv4-0"
I0126 15:20:19.091664       1 csi.go:188] "Persistent volume had no name for claim" PVC="default/pv-data4-stspv4-0"
I0126 15:20:19.091692       1 scheduler_binder.go:266] FindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.104"
I0126 15:20:19.091709       1 scheduler_binder.go:266] FindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.105"
I0126 15:20:19.091716       1 scheduler_binder.go:856] No matching volumes for Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0" on node "cn-hangzhou.10.0.0.104"
I0126 15:20:19.091734       1 scheduler_binder.go:856] No matching volumes for Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0" on node "cn-hangzhou.10.0.0.105"
I0126 15:20:19.091754       1 scheduler_binder.go:920] Provisioning for 1 claims of pod "default/stspv4-0" that has no matching volumes on node "cn-hangzhou.10.0.0.104" ...
I0126 15:20:19.091664       1 csi.go:188] "Persistent volume had no name for claim" PVC="default/pv-data4-stspv4-0"
I0126 15:20:19.091795       1 scheduler_binder.go:266] FindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.103"
I0126 15:20:19.091808       1 scheduler_binder.go:856] No matching volumes for Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0" on node "cn-hangzhou.10.0.0.103"
I0126 15:20:19.091823       1 scheduler_binder.go:920] Provisioning for 1 claims of pod "default/stspv4-0" that has no matching volumes on node "cn-hangzhou.10.0.0.103" ...
I0126 15:20:19.091763       1 scheduler_binder.go:920] Provisioning for 1 claims of pod "default/stspv4-0" that has no matching volumes on node "cn-hangzhou.10.0.0.105" ...
I0126 15:20:19.092116       1 scheduler_binder.go:378] AssumePodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.103"
I0126 15:20:19.092128       1 scheduler_binder.go:742] PVC "default/pv-data4-stspv4-0" is not bound
I0126 15:20:19.092149       1 scheduler_assume_cache.go:328] Assumed v1.PersistentVolumeClaim "default/pv-data4-stspv4-0", version 34784
I0126 15:20:19.092232       1 volume_binding.go:329] "Trying to bind volumes for pod" pod="default/stspv4-0"
I0126 15:20:19.092243       1 scheduler_binder.go:451] BindPodVolumes for pod "default/stspv4-0", node "cn-hangzhou.10.0.0.103"
I0126 15:20:19.092254       1 scheduler_binder.go:537] bindAPIUpdate: Pod "default/stspv4-0", PVC "default/pv-data4-stspv4-0"
I0126 15:20:19.099687       1 round_trippers.go:454] PUT https://10.0.0.100:6443/api/v1/namespaces/default/persistentvolumeclaims/pv-data4-stspv4-0 200 OK in 7 milliseconds
I0126 15:20:19.355211       1 round_trippers.go:454] GET https://10.0.0.100:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=5s 200 OK in 5 milliseconds
I0126 15:20:19.359628       1 round_trippers.go:454] PUT https://10.0.0.100:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=5s 200 OK in 4 milliseconds
I0126 15:20:19.359728       1 leaderelection.go:278] successfully renewed lease kube-system/kube-scheduler
I0126 15:20:19.553408       1 cache.go:732] Couldn't expire cache for pod default/stspv4-0. Binding is still in progress.
I0126 15:20:20.100639       1 volume_binding.go:332] "Failed to bind volumes for pod" pod="default/stspv4-0" err="binding volumes: provisioning failed for PVC \"pv-data4-stspv4-0\""
E0126 15:20:20.100680       1 framework.go:863] "Failed running PreBind plugin" err="binding volumes: provisioning failed for PVC \"pv-data4-stspv4-0\"" plugin="VolumeBinding" pod="default/stspv4-0"
I0126 15:20:20.100710       1 scheduler_assume_cache.go:342] Restored v1.PersistentVolumeClaim "default/pv-data4-stspv4-0"
E0126 15:20:20.100774       1 factory.go:397] "Error scheduling pod; retrying" err="running PreBind plugin \"VolumeBinding\": binding volumes: provisioning failed for PVC \"pv-data4-stspv4-0\"" pod="default/stspv4-0"
I0126 15:20:20.100874       1 scheduler.go:414] "Updating pod condition" pod="default/stspv4-0" conditionType=PodScheduled conditionStatus=False conditionReason="SchedulerError"
I0126 15:20:20.107673       1 round_trippers.go:454] POST https://10.0.0.100:6443/apis/events.k8s.io/v1/namespaces/default/events 201 Created in 6 milliseconds
I0126 15:20:21.366085       1 round_trippers.go:454] GET https://10.0.0.100:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=5s 200 OK in 5 milliseconds
I0126 15:20:21.371759       1 round_trippers.go:454] PUT https://10.0.0.100:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=5s 200 OK in 5 milliseconds
```

Is there anything i missed ? The request storage size is only available on node   `cn-hangzhou.10.0.0.105` but it seems the `scheduler` doesn't know it.

**Comment by pohly** [comment]
Good report, thanks for all the details.

The key question is indeed why the scheduler doesn't check `CSIStorageCapacity`. Is this the upstream kube-scheduler or some modified scheduler? As far as I know, Alibaba is using a custom scheduler in their managed Kubernetes. Is the CSIStorageCapacity feature enabled there?

Here's the code which checks for capacity:
https://github.com/kubernetes/kubernetes/blob/25697c4812daa06670efb97c49db7646502ed256/pkg/scheduler/framework/plugins/volumebinding/binder.go#L922-L974

However, it's intentionally fairly silent in cases where it accepts a node. I didn't want to emit too many log messages for the normal case. Perhaps `klog.V(5)` (= debug level) messages should be added for some of the `return` statements.

**Comment by mrpre** [comment]
@pohly Thanks for your guidance. After I read the implement of `hasEnoughCapacity ` 
https://github.com/kubernetes/kubernetes/blob/de17b6fef92b8525bf05656672db96b5607499b1/pkg/controller/volume/scheduling/scheduler_binder.go#L969-L982

I find all the `csistoragecapacities` the `external-provisioner` created for each node would be add to comparison and the key problem is that I set the same `Topology` for each `NodeGetInfo` interface, which make each `csistoragecapacitiy` is accessible in comparison.   

After I set the nodeid as the `Topology` for each `NodeGetInfo`, * the problem is solved *.  

It seems that `Topology` is not ZONE related feature if  `external-provisioner` is runed with "--node-deployment" using `Daemonset`. Otherwise the scheduler need more logic.
